project_material/tests/test_task_material.py

```diff
@@ -0,0 +1,358 @@
+
+    def test_if_material_line_deleted__stock_move_cancelled(self):
+        line = self._create_material_line()
+        move = line.move_ids
+        line.unlink()
+        assert move.state == 'cancel'
+
+    def test_if_any_move_done__material_line_can_not_be_deleted(self):
+        line = self._create_material_line(initial_qty=10)
+        self._force_transfer_move(line.move_ids, 1)
+        with pytest.raises(ValidationError):
+            line.unlink()
+
+    def _get_po_line(self, product):
+        return self.env['purchase.order.line'].search([('product_id', '=', product.id)])
+
+    def test_if_product_is_not_mto__purchase_order_not_generated(self):
+        self._create_material_line()
+        po_line = self._get_po_line(self.product_a)
+        assert not po_line
+
+    @data('one_step', 'two_steps')
+    def test_if_product_is_mto__purchase_order_generated(self, steps):
+        self.warehouse.consu_steps = steps
+        self.product_a.route_ids |= self.env.ref('stock.route_warehouse0_mto')
+        self._create_material_line()
+        po_line = self._get_po_line(self.product_a)
+        assert po_line
+
+    def test_after_change_task__initial_stock_move_cancelled(self):
+        line = self._create_material_line()
+        move = line.move_ids
+        line.task_id = self.task_2
+        assert move.state == 'cancel'
+
+    def test_after_change_task__new_stock_move_created(self):
+        line = self._create_material_line()
+        initial_move = line.move_ids
+        assert len(initial_move) == 1
+
+        line.task_id = self.task_2
+
+        new_move = line.move_ids
+        assert len(new_move) == 1
+        assert new_move != initial_move
+
+    def test_after_change_product__initial_stock_move_cancelled(self):
+        line = self._create_material_line()
+        move = line.move_ids
+        line.product_id = self.product_b
+        assert move.state == 'cancel'
+
+    def test_after_change_product__new_stock_move_created(self):
+        line = self._create_material_line()
+        initial_move = line.move_ids
+        assert len(initial_move) == 1
+
+        line.product_id = self.product_b
+
+        new_move = line.move_ids
+        assert len(new_move) == 1
+        assert new_move != initial_move
+
+    def test_if_any_move_done__material_line_can_not_change_task(self):
+        line = self._create_material_line(initial_qty=10)
+        self._force_transfer_move(line.move_ids, 1)
+        with pytest.raises(ValidationError):
+            line.task_id = self.task_2
+
+    def test_if_any_move_done__material_line_can_not_change_product(self):
+        line = self._create_material_line(initial_qty=10)
+        self._force_transfer_move(line.move_ids, 1)
+        with pytest.raises(ValidationError):
+            line.product_id = self.product_b
+
+    def test_for_each_material_line__one_stock_move_generated(self):
+        """Test each material line generates an independant stock move.
```

Comment:
Codacy found an issue: [Multi-line docstring summary should start at the second line (D213)](https://app.codacy.com/gh/Numigi/odoo-project-addons/pullRequest?prid=6165869)